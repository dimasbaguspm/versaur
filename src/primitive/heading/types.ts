import type { HTMLAttributes, ReactNode } from 'react'

/**
 * HeadingProps defines the props for the Heading component
 * @property level - Heading level (1-6, corresponding to h1-h6)
 * @property color - Versaur color system (primary, secondary, tertiary, ghost, neutral, success, info, warning, danger)
 * @property hasUnderline - Whether to underline the text
 * @property isCapitalize - Whether to capitalize the text
 * @property hasMargin - Whether to add margin bottom (mb-4)
 * @property align - Text alignment
 * @property italic - Whether to italicize the text
 * @property clamp - Clamp lines (1-5) or none
 * @property ellipsis - Whether to truncate text with ellipsis
 * @property className - Additional CSS classes
 * @property children - Text content
 */
export interface HeadingProps extends HTMLAttributes<HTMLHeadingElement> {
  /** Heading level (1-6, corresponding to h1-h6) */
  level?: 1 | 2 | 3 | 4 | 5 | 6
  /** Versaur color system */
  color?:
    | 'primary'
    | 'secondary'
    | 'tertiary'
    | 'ghost'
    | 'neutral'
    | 'success'
    | 'info'
    | 'warning'
    | 'danger'
    | 'inherit'
    | 'white'
    | 'black'
    | 'gray'
  /** Underline text */
  hasUnderline?: boolean
  /** Capitalize text */
  isCapitalize?: boolean
  /** Add margin bottom (mb-4) */
  hasMargin?: boolean
  /** Text alignment */
  align?: 'left' | 'center' | 'right' | 'justify'
  /** Italic text */
  italic?: boolean
  /** Clamp lines (1-5) or none */
  clamp?: 1 | 2 | 3 | 4 | 5 | 'none'
  /** Ellipsis (truncate) */
  ellipsis?: boolean
  /** Text content */
  children: ReactNode
}
