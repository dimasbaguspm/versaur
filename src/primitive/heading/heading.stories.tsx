import type { Meta, StoryObj } from '@storybook/react'
import { Heading } from './heading'
import type { HeadingProps } from './types'

const meta: Meta<typeof Heading> = {
  title: 'Primitive/Heading',
  component: Heading,
  tags: ['autodocs'],
  argTypes: {
    level: {
      control: 'select',
      options: [1, 2, 3, 4, 5, 6],
      description: 'Heading level (1-6, corresponding to h1-h6)',
      table: { type: { summary: '1 | 2 | 3 | 4 | 5 | 6' } },
    },
    color: {
      control: 'select',
      options: [
        'primary',
        'secondary',
        'tertiary',
        'ghost',
        'neutral',
        'success',
        'info',
        'warning',
        'danger',
        'inherit',
        'white',
        'black',
        'gray',
      ],
      description: 'Versaur color system',
      table: { type: { summary: 'Color' } },
    },
    hasUnderline: {
      control: 'boolean',
      description: 'Underline text',
      table: { type: { summary: 'boolean' } },
    },
    isCapitalize: {
      control: 'boolean',
      description: 'Capitalize text',
      table: { type: { summary: 'boolean' } },
    },
    hasMargin: {
      control: 'boolean',
      description: 'Add margin bottom (mb-4)',
      table: { type: { summary: 'boolean' } },
    },
    align: {
      control: 'select',
      options: ['left', 'center', 'right', 'justify'],
      description: 'Text alignment',
      table: { type: { summary: 'left | center | right | justify' } },
    },
    italic: {
      control: 'boolean',
      description: 'Italic text',
      table: { type: { summary: 'boolean' } },
    },
    clamp: {
      control: 'select',
      options: [1, 2, 3, 4, 5, 'none'],
      description: 'Clamp lines (1-5)',
      table: { type: { summary: '1 | 2 | 3 | 4 | 5 | none' } },
    },
    ellipsis: {
      control: 'boolean',
      description: 'Truncate text with ellipsis',
      table: { type: { summary: 'boolean' } },
    },
  },
} satisfies Meta<typeof Heading>

export default meta

type Story = StoryObj<HeadingProps>

/**
 * Default heading component showcasing h1 with primary color
 */
export const Default: Story = {
  args: {
    level: 1,
    color: 'ghost',
    children: 'Default Heading',
  },
}

/**
 * Primary heading with all features enabled
 */
export const PrimaryWithFeatures: Story = {
  args: {
    level: 1,
    color: 'primary',
    hasUnderline: true,
    isCapitalize: true,
    hasMargin: true,
    align: 'center',
    italic: true,
    children: 'Primary Heading: Underline, Capitalize, Center, Italic, Margin',
  },
}

/**
 * Secondary heading with right alignment and clamp
 */
export const SecondaryWithClamp: Story = {
  args: {
    level: 2,
    color: 'secondary',
    align: 'right',
    clamp: 2,
    hasMargin: true,
    children:
      'Secondary Heading: Right aligned with line clamp set to 2. This is a very long text that should be clamped at 2 lines when the container is narrow enough.',
  },
}

/**
 * Tertiary heading with ellipsis and underline
 */
export const TertiaryWithEllipsis: Story = {
  args: {
    level: 3,
    color: 'tertiary',
    ellipsis: true,
    hasUnderline: true,
    hasMargin: true,
    children:
      'Tertiary Heading: This is a very long text that will be truncated with ellipsis',
  },
}

/**
 * Success heading with level 4
 */
export const SuccessHeading: Story = {
  args: {
    level: 4,
    color: 'success',
    hasMargin: true,
    children: 'Success Heading',
  },
}

/**
 * Warning heading with capitalize and justify alignment
 */
export const WarningCapitalize: Story = {
  args: {
    level: 5,
    color: 'warning',
    isCapitalize: true,
    align: 'justify',
    hasMargin: true,
    children: 'warning heading with capitalize and justify alignment',
  },
}

/**
 * Danger heading with italic and underline
 */
export const DangerItalicUnderline: Story = {
  args: {
    level: 6,
    color: 'danger',
    italic: true,
    hasUnderline: true,
    hasMargin: true,
    children: 'Danger Heading: Italic and Underlined',
  },
}

/**
 * All heading levels showcase
 */
export const AllLevels: Story = {
  render: () => (
    <div className='space-y-4'>
      <Heading level={1} color='primary' hasMargin>
        H1 Heading - Primary Color
      </Heading>
      <Heading level={2} color='secondary' hasMargin>
        H2 Heading - Secondary Color
      </Heading>
      <Heading level={3} color='tertiary' hasMargin>
        H3 Heading - Tertiary Color
      </Heading>
      <Heading level={4} color='success' hasMargin>
        H4 Heading - Success Color
      </Heading>
      <Heading level={5} color='warning' hasMargin>
        H5 Heading - Warning Color
      </Heading>
      <Heading level={6} color='danger' hasMargin>
        H6 Heading - Danger Color
      </Heading>
    </div>
  ),
}

/**
 * Color variants showcase
 */
export const ColorVariants: Story = {
  render: () => (
    <div className='space-y-4'>
      <Heading level={3} color='primary' hasMargin>
        Primary Heading
      </Heading>
      <Heading level={3} color='secondary' hasMargin>
        Secondary Heading
      </Heading>
      <Heading level={3} color='tertiary' hasMargin>
        Tertiary Heading
      </Heading>
      <Heading level={3} color='ghost' hasMargin>
        Ghost Heading
      </Heading>
      <Heading level={3} color='neutral' hasMargin>
        Neutral Heading
      </Heading>
      <Heading level={3} color='success' hasMargin>
        Success Heading
      </Heading>
      <Heading level={3} color='info' hasMargin>
        Info Heading
      </Heading>
      <Heading level={3} color='warning' hasMargin>
        Warning Heading
      </Heading>
      <Heading level={3} color='danger' hasMargin>
        Danger Heading
      </Heading>
    </div>
  ),
}

/**
 * Alignment variants showcase
 */
export const AlignmentVariants: Story = {
  render: () => (
    <div className='space-y-4'>
      <Heading level={3} align='left' hasMargin>
        Left Aligned Heading
      </Heading>
      <Heading level={3} align='center' hasMargin>
        Center Aligned Heading
      </Heading>
      <Heading level={3} align='right' hasMargin>
        Right Aligned Heading
      </Heading>
      <Heading level={3} align='justify' hasMargin>
        Justify Aligned Heading - This is a longer heading text to demonstrate
        justify alignment behavior
      </Heading>
    </div>
  ),
}
