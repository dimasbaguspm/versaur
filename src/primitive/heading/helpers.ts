import { cva } from '@/utils/variants'

/**
 * Heading variants for Versaur design system
 * Supports color, underline, capitalization, margin, alignment, italic, clamp, ellipsis, and heading levels
 */
export const headingVariants = cva('', {
  variants: {
    color: {
      primary: 'text-primary',
      secondary: 'text-secondary',
      tertiary: 'text-tertiary',
      ghost: 'text-ghost',
      neutral: 'text-ghost',
      success: 'text-success',
      info: 'text-info',
      warning: 'text-warning',
      danger: 'text-danger',
      inherit: '',
      gray: 'text-gray-500',
      black: 'text-black',
      white: 'text-white',
    },
    hasUnderline: {
      true: 'underline',
      false: '',
    },
    isCapitalize: {
      true: 'capitalize',
      false: '',
    },
    hasMargin: {
      true: 'mb-4',
      false: '',
    },
    align: {
      left: 'text-left',
      center: 'text-center',
      right: 'text-right',
      justify: 'text-justify',
    },
    italic: {
      true: 'italic',
      false: '',
    },
    clamp: {
      1: 'line-clamp-1',
      2: 'line-clamp-2',
      3: 'line-clamp-3',
      4: 'line-clamp-4',
      5: 'line-clamp-5',
      none: '',
    },
    ellipsis: {
      true: 'truncate',
      false: '',
    },
    level: {
      1: 'font-bold text-2xl leading-loose',
      2: 'font-semibold text-xl leading-relaxed',
      3: 'font-medium text-base leading-relaxed',
      4: 'font-bold text-sm leading-normal',
      5: 'font-semibold text-sm leading-normal',
      6: 'font-medium text-xs leading-normal',
    },
  },
  defaultVariants: {
    color: 'ghost',
    hasUnderline: false,
    isCapitalize: false,
    hasMargin: false,
    align: 'left',
    italic: false,
    clamp: 'none',
    ellipsis: false,
    level: 1,
  },
})
