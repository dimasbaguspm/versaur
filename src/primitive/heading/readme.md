# Heading Component

A flexible and accessible heading component for Versaur UI that renders semantic HTML headings
(h1-h6) with comprehensive styling options.

## Features

- **Semantic HTML**: Renders proper heading tags (h1-h6) based on the `level` prop
- **Color System**: Supports all Versaur color variants (primary, secondary, tertiary, ghost,
  neutral, success, info, warning, danger)
- **Typography Control**: Built-in typography styles with optional font size and weight overrides
- **Text Styling**: Underline, capitalization, italics, alignment options
- **Layout Control**: Optional margin bottom with `hasMargin` prop
- **Text Overflow**: Line clamping and ellipsis truncation support
- **Accessibility**: Follows semantic heading structure for screen readers

## Usage

### Basic Usage

```tsx
import { Heading } from '@versaur/ui'

// Simple h1 heading
<Heading level={1}>Main Title</Heading>

// h2 with primary color and margin
<Heading level={2} color="primary" hasMargin>
  Section Title
</Heading>

// h3 with multiple features
<Heading
  level={3}
  color="secondary"
  hasUnderline
  isCapitalize
  align="center"
>
  Styled Heading
</Heading>
```

### Typography Overrides

```tsx
// Override default font size and weight
<Heading level={4} fontSize="6xl" fontWeight="light">
  Custom Sized Heading
</Heading>

// Small h1 for special cases
<Heading level={1} fontSize="sm" fontWeight="normal">
  Compact Main Title
</Heading>
```

### Text Layout and Overflow

```tsx
// Centered heading with margin
<Heading level={2} align="center" hasMargin>
  Centered Title
</Heading>

// Truncated heading with ellipsis
<Heading level={3} ellipsis>
  This is a very long heading that will be truncated
</Heading>

// Multi-line heading with line clamping
<Heading level={4} clamp={2}>
  This is a long heading that will be clamped to exactly two lines
</Heading>
```

### Color Variants

```tsx
<Heading level={2} color="success">Success Heading</Heading>
<Heading level={2} color="warning">Warning Heading</Heading>
<Heading level={2} color="danger">Error Heading</Heading>
<Heading level={2} color="info">Info Heading</Heading>
```

## Props

| Prop           | Type                                         | Default     | Description                          |
| -------------- | -------------------------------------------- | ----------- | ------------------------------------ |
| `level`        | `1 \| 2 \| 3 \| 4 \| 5 \| 6`                 | `1`         | Heading level (corresponds to h1-h6) |
| `color`        | `Color`                                      | `'ghost'`   | Color from Versaur color system      |
| `hasUnderline` | `boolean`                                    | `false`     | Add underline decoration             |
| `isCapitalize` | `boolean`                                    | `false`     | Capitalize text                      |
| `hasMargin`    | `boolean`                                    | `false`     | Add margin bottom (mb-4)             |
| `align`        | `'left' \| 'center' \| 'right' \| 'justify'` | `'left'`    | Text alignment                       |
| `italic`       | `boolean`                                    | `false`     | Italic text                          |
| `clamp`        | `1 \| 2 \| 3 \| 4 \| 5 \| 'none'`            | `'none'`    | Line clamping                        |
| `ellipsis`     | `boolean`                                    | `false`     | Truncate with ellipsis               |
| `fontSize`     | `FontSize`                                   | `undefined` | Override default font size           |
| `fontWeight`   | `FontWeight`                                 | `undefined` | Override default font weight         |
| `className`    | `string`                                     | `undefined` | Additional CSS classes               |
| `children`     | `ReactNode`                                  | -           | Heading content                      |

## Default Typography Scale

| Level | Tag  | Font Weight     | Font Size   | Line Height       |
| ----- | ---- | --------------- | ----------- | ----------------- |
| 1     | `h1` | `font-bold`     | `text-4xl`  | `leading-loose`   |
| 2     | `h2` | `font-semibold` | `text-3xl`  | `leading-relaxed` |
| 3     | `h3` | `font-medium`   | `text-2xl`  | `leading-relaxed` |
| 4     | `h4` | `font-bold`     | `text-xl`   | `leading-normal`  |
| 5     | `h5` | `font-semibold` | `text-lg`   | `leading-normal`  |
| 6     | `h6` | `font-medium`   | `text-base` | `leading-normal`  |

## Accessibility

- Uses semantic HTML heading tags (h1-h6) for proper document structure
- Maintains logical heading hierarchy for screen readers
- Supports all standard HTML heading attributes
- Color contrast meets WCAG 2.1 AA standards

## Best Practices

1. **Heading Hierarchy**: Use heading levels sequentially (h1 → h2 → h3, etc.)
2. **Unique H1**: Only use one h1 per page for the main title
3. **Descriptive Content**: Make heading text descriptive and meaningful
4. **Consistent Spacing**: Use `hasMargin` consistently for layout spacing
5. **Color Meaning**: Use semantic colors (success, warning, danger) meaningfully
