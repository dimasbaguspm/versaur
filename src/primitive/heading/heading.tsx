import { forwardRef } from 'react'
import { cn } from '@/utils/cn'
import { headingVariants } from './helpers'
import type { HeadingProps } from './types'

/**
 * Heading component for Versaur UI
 * Provides semantic heading typography (h1-h6), color, underline, capitalization, and margin support
 * @example <Heading level={1} color="primary" hasUnderline isCapitalize hasMargin>Main Title</Heading>
 */
export const Heading = forwardRef<HTMLHeadingElement, HeadingProps>(
  (
    {
      level = 3,
      color = 'black',
      hasUnderline = false,
      isCapitalize = false,
      hasMargin = false,
      align = 'left',
      italic = false,
      clamp = 'none',
      ellipsis = false,
      className,
      children,
      ...props
    },
    ref
  ) => {
    // Determine the HTML tag based on level
    const HeadingTag = `h${level}` as 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'

    return (
      <HeadingTag
        ref={ref}
        className={cn(
          headingVariants({
            color,
            hasUnderline,
            isCapitalize,
            hasMargin,
            align,
            italic,
            clamp,
            ellipsis,
            level,
          }),
          className
        )}
        {...props}
      >
        {children}
      </HeadingTag>
    )
  }
)

Heading.displayName = 'Heading'
