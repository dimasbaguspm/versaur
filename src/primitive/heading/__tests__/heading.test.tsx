import { render } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../heading.stories'

describe('Heading', () => {
  const {
    Default,
    PrimaryWithFeatures,
    SecondaryWithClamp,
    TertiaryWithEllipsis,
    SuccessHeading,
    WarningCapitalize,
    DangerItalicUnderline,
    AllLevels,
    ColorVariants,
    AlignmentVariants,
  } = composeStories(stories)

  it('renders default heading correctly', () => {
    const { getByText } = render(<Default />)
    const el = getByText('Default Heading')
    expect(el.tagName).toBe('H1')
    expect(el.className).toContain('font-bold')
    expect(el.className).toContain('text-2xl')
    expect(el.className).toContain('leading-loose')
    expect(el.className).toContain('text-ghost')
  })

  it('renders primary heading with all features', () => {
    const { getByText } = render(<PrimaryWithFeatures />)
    const el = getByText(
      /Primary Heading: Underline, Capitalize, Center, Italic, Margin/i
    )
    expect(el.tagName).toBe('H1')
    expect(el.className).toContain('font-bold')
    expect(el.className).toContain('text-2xl')
    expect(el.className).toContain('leading-loose')
    expect(el.className).toContain('text-primary')
    expect(el.className).toContain('underline')
    expect(el.className).toContain('capitalize')
    expect(el.className).toContain('text-center')
    expect(el.className).toContain('italic')
    expect(el.className).toContain('mb-4')
  })

  it('renders secondary heading with clamp and right alignment', () => {
    const { getByText } = render(<SecondaryWithClamp />)
    const el = getByText(/Secondary Heading: Right aligned with line clamp/i)
    expect(el.tagName).toBe('H2')
    expect(el.className).toContain('font-semibold')
    expect(el.className).toContain('text-xl')
    expect(el.className).toContain('leading-relaxed')
    expect(el.className).toContain('text-secondary')
    expect(el.className).toContain('text-right')
    expect(el.className).toContain('line-clamp-2')
    expect(el.className).toContain('mb-4')
  })

  it('renders tertiary heading with ellipsis and underline', () => {
    const { getByText } = render(<TertiaryWithEllipsis />)
    const el = getByText(/Tertiary Heading: This is a very long text/i)
    expect(el.tagName).toBe('H3')
    expect(el.className).toContain('font-semibold')
    expect(el.className).toContain('text-lg')
    expect(el.className).toContain('leading-relaxed')
    expect(el.className).toContain('text-tertiary')
    expect(el.className).toContain('truncate')
    expect(el.className).toContain('underline')
    expect(el.className).toContain('mb-4')
  })

  it('renders success heading correctly', () => {
    const { getByText } = render(<SuccessHeading />)
    const el = getByText(/Success Heading/i)
    expect(el.tagName).toBe('H4')
    expect(el.className).toContain('font-bold')
    expect(el.className).toContain('text-sm')
    expect(el.className).toContain('leading-normal')
    expect(el.className).toContain('text-success')
    expect(el.className).toContain('mb-4')
  })

  it('renders warning heading with capitalize and justify alignment', () => {
    const { getByText } = render(<WarningCapitalize />)
    const el = getByText(/warning heading with capitalize and justify/i)
    expect(el.tagName).toBe('H5')
    expect(el.className).toContain('font-semibold')
    expect(el.className).toContain('text-sm')
    expect(el.className).toContain('leading-normal')
    expect(el.className).toContain('text-warning')
    expect(el.className).toContain('capitalize')
    expect(el.className).toContain('text-justify')
    expect(el.className).toContain('mb-4')
  })

  it('renders danger heading with italic and underline', () => {
    const { getByText } = render(<DangerItalicUnderline />)
    const el = getByText(/Danger Heading: Italic and Underlined/i)
    expect(el.tagName).toBe('H6')
    expect(el.className).toContain('font-medium')
    expect(el.className).toContain('text-xs')
    expect(el.className).toContain('leading-normal')
    expect(el.className).toContain('text-danger')
    expect(el.className).toContain('italic')
    expect(el.className).toContain('underline')
    expect(el.className).toContain('mb-4')
  })

  it('renders all heading levels correctly', () => {
    const { container } = render(<AllLevels />)
    const headings = container.querySelectorAll('h1, h2, h3, h4, h5, h6')
    expect(headings).toHaveLength(6)

    // Check each heading level
    expect(container.querySelector('h1')).toHaveTextContent(
      'H1 Heading - Primary Color'
    )
    expect(container.querySelector('h2')).toHaveTextContent(
      'H2 Heading - Secondary Color'
    )
    expect(container.querySelector('h3')).toHaveTextContent(
      'H3 Heading - Tertiary Color'
    )
    expect(container.querySelector('h4')).toHaveTextContent(
      'H4 Heading - Success Color'
    )
    expect(container.querySelector('h5')).toHaveTextContent(
      'H5 Heading - Warning Color'
    )
    expect(container.querySelector('h6')).toHaveTextContent(
      'H6 Heading - Danger Color'
    )
  })

  it('renders color variants correctly', () => {
    const { container } = render(<ColorVariants />)
    const headings = container.querySelectorAll('h3')
    expect(headings).toHaveLength(9)

    // Check color classes
    expect(headings[0]).toHaveClass('text-primary')
    expect(headings[1]).toHaveClass('text-secondary')
    expect(headings[2]).toHaveClass('text-tertiary')
    expect(headings[3]).toHaveClass('text-ghost')
    expect(headings[4]).toHaveClass('text-ghost') // neutral maps to ghost
    expect(headings[5]).toHaveClass('text-success')
    expect(headings[6]).toHaveClass('text-info')
    expect(headings[7]).toHaveClass('text-warning')
    expect(headings[8]).toHaveClass('text-danger')
  })

  it('renders alignment variants correctly', () => {
    const { container } = render(<AlignmentVariants />)
    const headings = container.querySelectorAll('h3')
    expect(headings).toHaveLength(4)

    // Check alignment classes
    expect(headings[0]).toHaveClass('text-left')
    expect(headings[1]).toHaveClass('text-center')
    expect(headings[2]).toHaveClass('text-right')
    expect(headings[3]).toHaveClass('text-justify')
  })

  it('matches snapshot', () => {
    expect(render(<Default />).asFragment()).toMatchSnapshot()
  })

  it('matches snapshot for primary with all features', () => {
    expect(render(<PrimaryWithFeatures />).asFragment()).toMatchSnapshot()
  })
})
