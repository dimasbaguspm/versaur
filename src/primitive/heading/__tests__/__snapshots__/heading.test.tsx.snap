// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Heading > matches snapshot 1`] = `
<DocumentFragment>
  <h1
    class="text-ghost text-left font-bold text-2xl leading-loose"
  >
    Default Heading
  </h1>
</DocumentFragment>
`;

exports[`Heading > matches snapshot for primary with all features 1`] = `
<DocumentFragment>
  <h1
    class="text-primary underline capitalize mb-4 text-center italic font-bold text-2xl leading-loose"
  >
    Primary Heading: Underline, Capitalize, Center, Italic, Margin
  </h1>
</DocumentFragment>
`;
